File: oscar-algo/src/main/java/oscar/algo/reversible/ReversibleSparseSetJava.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.algo.reversible;


import java.util.Arrays;



/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class ReversibleSparseSetJava extends AbstractSparseSet {
	

	private ReversibleInt size;
	private ReversibleInt maxV;
	private ReversibleInt minV;

	private ReversibleContext s;
	
	/**
	 * Initializes a set with all values min..max in it
	 * @param s
	 * @param min
	 * @param max >= min
	 */
	public ReversibleSparseSetJava(ReversibleContext s, int min, int max) {
		this(s,min,max,false);
	}
	
	/**
	 * Initializes a set that can potentially contain all values min..max in it
	 * @param s
	 * @param min
	 * @param max >= min
	 */
	public ReversibleSparseSetJava(ReversibleContext s, int min, int max, boolean empty) {
		this.s = s;
		initIndexes(min, max, empty);
	}

	@Override
	protected void createSizeMinMax() {
		size = new ReversibleInt(s, 0);
		minV = new ReversibleInt(s, 0);
		maxV = new ReversibleInt(s, 0);
	}

	@Override
	protected void setSize(int size) {
		this.size.setValue(size);
	}

	@Override
	protected void setMin(int min) {
		minV.setValue(min);
	}

	@Override
	protected void setMax(int max) {
		maxV.setValue(max);
	}

	@Override
	public int getSize() {
		return size.getValue();
	}

	@Override
	public int getMin() {
		return minV.getValue();
	}

	@Override
	public int getMax() {
		return maxV.getValue();
	}

    public String toString() {
        return Arrays.toString(getSortedVals());
    }
	
}

File: src/main/scala/oscar/algo/reversible/ReversibleSparseSetJava.scala
package oscar.algo.reversible

import java.util.Arrays

class ReversibleSparseSetJava(val s: ReversibleContext, min: Int, max: Int, empty: Boolean = false) extends AbstractSparseSet {
  def this(s: ReversibleContext, min: Int, max: Int) = this(s, min, max, false)
  private var _size: ReversibleInt = _
  private var maxV: ReversibleInt = _
  private var minV: ReversibleInt = _

  initIndexes(min, max, empty)

  override protected def createSizeMinMax(): Unit = {
    _size = new ReversibleInt(s, 0)
    minV = new ReversibleInt(s, 0)
    maxV = new ReversibleInt(s, 0)
  }

  override protected def setSize(size: Int): Unit = {
    this._size.setValue(size)
  }

  override protected def setMin(min: Int): Unit = {
    minV.setValue(min)
  }

  override protected def setMax(max: Int): Unit = {
    maxV.setValue(max)
  }

  override def getSize(): Int = _size.getValue()

  override def getMin(): Int = minV.getValue()

  override def getMax(): Int = maxV.getValue()

  override def toString: String = {
    Arrays.toString(getSortedVals)
  }
}
